crates/derive/src/combine_api.rs

```diff
@@ -225,11 +269,18 @@ pub fn drain_chunk(
     arena: &mut Vec<u8>,
     out: &mut Vec<cgo::Out>,
     stats: &mut DocCounter,
+    shape: &mut Option<&mut doc::inference::Shape>,
 ) -> Result<bool, doc::combine::Error> {
     // Convert target from a delta to an absolute target length of the arena.
     let target_length = target_length + arena.len();
 
     drainer.drain_while(|doc, fully_reduced| {
+        if let Some(shape) = shape.as_mut() {
+            match &doc {
+                doc::LazyNode::Node(n) => shape.widen(*n, Location::Root),
```

Comment:
you need to call the limit function if this returns true -- a single combiner can have a _lot_ of documents.

Reply:
🤦 yup